Sancov guard draft

ramosian-glider opened the pull request.

ramosian-glider committed:
kcov: factor out struct kcov_state

Group several kcov-related fields (area, size, mode, sequence) that
are stored in various structures, into `struct kcov_state`, so that
these fields can be easily passed around and manipulated.

This prepares us for the upcoming change that will introduce more
kcov state.

Signed-off-by: Alexander Potapenko <glider@google.com>

ramosian-glider committed:
kcov: apply clang-format to kernel/kcov.c and include/linux/kcov.h

kcov used to obey clang-format style, but somehow diverted over time.
No functional change.

Signed-off-by: Alexander Potapenko <glider@google.com>

ramosian-glider committed:
kcov: x86: introduce CONFIG_KCOV_ENABLE_GUARDS

The new config switches coverage instrumentation to using
  __sanitizer_cov_trace_pc_guard(u32 *guard)
instead of
  __sanitizer_cov_trace_pc(void)

Each callback receives a unique 32-bit guard variable residing in the
__sancov_guards section. Those guards can be used by kcov to deduplicate
the coverage on the fly.

As a first step, we make the new instrumentation mode 1:1 compatible with
the old one.

Signed-off-by: Alexander Potapenko <glider@google.com>

ramosian-glider committed:
kcov: add ioctl(KCOV_UNIQUE_ENABLE)

ioctl(KCOV_UNIQUE_ENABLE) enables collection of deduplicated coverage
in the presence of CONFIG_KCOV_ENABLE_GUARDS.

The buffer shared with the userspace is divided in two parts, one holding
a bitmap, and the other one being the trace. The single parameter of
ioctl(KCOV_UNIQUE_ENABLE) determines the number of words used for the bitmap.

Each __sanitizer_cov_trace_pc_guard() instrumentation hook receives a
pointer to a unique guard variable. Upon the first call of each hook,
the guard variable is initialized with a unique integer, which is used
to map those hooks to bits in the bitmap. In the new coverage collection mode,
the kernel first checks whether the bit corresponding to a particular hook
is set, and then, if it is not, the PC is written into the trace buffer,
and the bit is set.

Also update the documentation.

Signed-off-by: Alexander Potapenko <glider@google.com>

jiangenj commented on kernel/kcov.c:
```diff
@@ -252,8 +273,8 @@ static void notrace write_comp_data(u64 type, u64 arg1, u64 arg2, u64 ip)
-		/* See comment in __sanitizer_cov_trace_pc(). */
-		WRITE_ONCE(area[0], count + 1);
+		/* See comment in sanitizer_cov_write_subsequent(). */
+		WRITE_ONCE(trace[0], count + 1);
```
trace is not defined yet in this patch: [kcov: x86: introduce CONFIG_KCOV_ENABLE_GUARDS](https://github.com/ramosian-glider/linux/pull/7/commits/bfd28cf7e2e838b62e453d752453d203b8781715)

ramosian-glider replied:
Good catch, fixed!

jiangenj commented:
Do you have the userspace prog to test now?
Also I doubt how sancov_guards for modules will be loaded.

ramosian-glider commented:
Here you go: 
[kcov.txt](https://github.com/user-attachments/files/18912693/kcov.txt)

(rename from kcov.txt to kcov.c)

As for modules, I haven't looked closely, but I don't expect any issues with those.

jiangenj commented on include/asm-generic/vmlinux.lds.h:
```diff
@@ -121,6 +121,17 @@ defined(CONFIG_AUTOFDO_CLANG) || defined(CONFIG_PROPELLER_CLANG)
+#if defined(CONFIG_KCOV_ENABLE_GUARDS)
+#define SANCOV_GUARDS_BSS 			\
+	__sancov_guards(NOLOAD) : {		\
+		__sancov_guards_start = .;	\
```
maybe we need different name for __sancov_guards_start.<module>?

ramosian-glider replied:
Why? I think it's common to declare SECTION_NAME_start and SECTION_NAME_end for SECTION_NAME.

quic-likaid commented on kernel/kcov.c:
```diff
@@ -229,14 +234,96 @@ void notrace __sanitizer_cov_trace_pc(void)
+ * For an unlikely case of a race with another task executing the same basic
+ * block, we keep a free index in a per-cpu variable.
+ * In an even less likely case a task can lose a race and get rescheduled onto a
+ * CPU that already has a saved index, discarding that index. This will result
```
If a task is running on CPU1, trying to initialize a guard with `saved_index` from the CPU, and after a successful write through `cmpxchg`, it gets rescheduled to another CPU. Then what about the `saved_index` on CPU1? It doesn't seem to be overwritten and possibly be reused to initialize other guards.

ramosian-glider replied:
You are right, we probably need a local_lock here.

ramosian-glider replied:
I think this is resolved now.

ramosian-glider committed:
kcov: add `trace` and `trace_size` to `struct kcov_state`

Keep kcov_state.area as the pointer to the memory buffer used by
kcov and shared with the userspace. Store the pointer to the trace
(part of the buffer holding sequential events) separately, as we will
be splitting that buffer in multiple parts.
No functional change so far.

Signed-off-by: Alexander Potapenko <glider@google.com>

ramosian-glider commented:
To simplify the API and do not require the user to unconditionally carve out the bitmap from the mmapped region, we decided to remove the buffer partitioning API and introduce ioctl(KCOV_UNIQUE_ENABLE).
Now it's possible to use the old code, but do one-off runs with deduplicated coverage if needed.

jiangenj commented on kernel/kcov.c:
```diff
@@ -552,6 +638,13 @@ static int kcov_get_mode(unsigned long arg)
+		return KCOV_MODE_TRACE_UNIQUE_PC;
+#else
+		return -ENOTSUPP;
+
```
extra empty line here

ramosian-glider replied:
Fixed.

jiangenj commented on kernel/kcov.c:
```diff
@@ -229,14 +234,93 @@ void notrace __sanitizer_cov_trace_pc(void)
+		this_cpu_write(saved_index, index);
+		return old_guard;
+	}
+	return index;
```
in previous version, userspace gets pos number as 177 and actual printed cover (not 0) is 22.
Curious to know in this version, how many actual covered pc / pos?
Please share the updated userspace program too.

ramosian-glider replied:
I am not sure I understand the question, can you please elaborate?

> Please share the updated userspace program too.

Please check the updated documentation, you basically need to replace KCOV_ENABLE with KCOV_UNIQUE_ENABLE and memset the bitmap with zeroes every time you want to reset the coverage.

ramosian-glider replied:
[kcov-new.txt](https://github.com/user-attachments/files/19109130/kcov-new.txt)

Here is the updated userspace program.

jiangenj replied:
For example, 
```
./kcov-guard
cover[0]: cea
cover[1]: ffffffff81c6472a
cover[2]: ffffffff81c647de
cover[3]: ffffffff81c6481e
cover[4]: ffffffff81c649a7
cover[5]: ffffffff81b54e8b
cover[6]: ffffffff81b4ee3b
cover[7]: ffffffff81b4eeb9
cover[8]: ffffffff81b4ef32
cover[9]: ffffffff81b4ef67
cover[10]: ffffffff81b54edf
cover[11]: ffffffff81b54f13
cover[12]: ffffffff81b54fcb
cover[13]: ffffffff823c3a04
cover[14]: ffffffff823c37aa
cover[15]: ffffffff82461107
cover[16]: ffffffff824612a0
cover[17]: ffffffff823c3869
cover[18]: ffffffff81c64cd9
cover[19]: ffffffff81c64d39
cover[20]: ffffffff81c63d8d
cover[21]: ffffffff81c63dea
cover[22]: ffffffff81883de0
cover[23]: ffffffff81883e8f
cover[24]: ffffffff81c6472a
cover[25]: ffffffff81c647de
cover[26]: ffffffff81c6481e
cover[27]: ffffffff81c649a7
cover[28]: ffffffff81b54e8b
cover[29]: ffffffff81b4ee3b
cover[30]: ffffffff81b4eeb9
cover[31]: ffffffff81b4ef32
cover[32]: ffffffff81b4ef67
cover[33]: ffffffff81b54edf
cover[34]: ffffffff81b54f13
cover[35]: ffffffff81b54fcb
cover[36]: ffffffff823c3a04
cover[37]: ffffffff823c37aa
cover[38]: ffffffff82461107
cover[39]: ffffffff824612a0
cover[40]: ffffffff823c3869
cover[41]: ffffffff81c64cd9
cover[42]: ffffffff81c64d39
cover[43]: ffffffff81c63d8d
cover[44]: ffffffff81c63dea
cover[45]: ffffffff81883de0
cover[46]: ffffffff81883e8f
```
cover[0] is the pos number which is 0xcea, while actually covered number is 46, so actual used index is only 46/0xcea =1.4%.

The same for uniq:
```
 ./kcov-guard uniq is 23/176=13.1%.
cover[0]: b0
cover[1]: ffffffff81c6472a
cover[2]: ffffffff81c647de
cover[3]: ffffffff81c6481e
cover[4]: ffffffff81c649a7
cover[5]: ffffffff81b54e8b
cover[6]: ffffffff81b4ee3b
cover[7]: ffffffff81b4eeb9
cover[8]: ffffffff81b4ef32
cover[9]: ffffffff81b4ef67
cover[10]: ffffffff81b54edf
cover[11]: ffffffff81b54f13
cover[12]: ffffffff81b54fcb
cover[13]: ffffffff823c3a04
cover[14]: ffffffff823c37aa
cover[15]: ffffffff82461107
cover[16]: ffffffff824612a0
cover[17]: ffffffff823c3869
cover[18]: ffffffff81c64cd9
cover[19]: ffffffff81c64d39
cover[20]: ffffffff81c63d8d
cover[21]: ffffffff81c63dea
cover[22]: ffffffff81883de0
cover[23]: ffffffff81883e8f
```

In sparse mode, it didn't print cover[0], cover[1];
```
./kcov-guard sparse
cover[2]: ffffffff81c64cd9
...
cover[176]: ffffffff82426910
cover[177]: ffffffff824242e6
```
After add printf for scan_size, it reports
```
scan_size=65536
```
So, it means overflow here. For 177 covered pc and kcov overflow happened.

ramosian-glider replied:
> cover[0] is the pos number which is 0xcea, while actually covered number is 46, so actual used index is only 46/0xcea =1.4%.

This happens because the coverage collection starts once we write the mode into `t->kcov_mode` when handling `ioctl(fd, KCOV_ENABLE)` and finishes once we write 0 into `t->kcov_mode` in `ioctl(fd, KCOV_DISABLE)`.

When running e.g.:
```
  ioctl(fd, KCOV_ENABLE, KCOV_TRACE_PC);
  read(-1, NULL, 0);
  ioctl(fd, KCOV_DISABLE, 0);
```

, the buffer will contain not only the coverage from the `read()` call, but also from whatever happened in the kernel while handling the ioctls as long as t->kcov_mode was non-zero.

To work around this, we reset `cover[0]` before executing the `read()` call, and save its value after that, so that the test program only prints the relevant coverage.

Note that this has always been the case, even the current kcov implementation requires this adjustment of `cover[0]`.

Naturally, for `KCOV_UNIQUE_ENABLE` the numbers of collected PCs are lower.

> In sparse mode, it didn't print cover[0], cover[1];

Right now kcov treats the whole buffer as a sparse array, i.e. it will write to `cover[*guard]` in `__sanitizer_cov_trace_pc_guard(u32 *guard)`. If the PCs that received the numbers 0 and 1 weren't executed, there will be zeroes in cover[0] and cover[1].

> So, it means overflow here. For 177 covered pc and kcov overflow happened.

It did not. In this mode, we have to scan the whole array, because we do not know where to stop.

I am actually not sure whether we need this mode at all. But if we decide to keep it, we probably need to reserve cover[0] for the size of the array to be scanned.

jiangenj commented:
> Here you go: [kcov.txt](https://github.com/user-attachments/files/18912693/kcov.txt)
> 
> (rename from kcov.txt to kcov.c)
> 
> As for modules, I haven't looked closely, but I don't expect any issues with those.

For example, If i load the module when trace_pc_guard enabled
```
 insmod ./usbmon.ko
[ 3686.608008][ T9399] module: usbmon: Unknown rela relocation: 42
insmod: ERROR: could not insert module ./usbmon.ko: Invalid module format
```

The readelf shows it has many __sancov_guards sections
```
...
  [294] __sancov_guards   PROGBITS         0000000000000000  00011ab8
  [295] __sancov_guards   PROGBITS         0000000000000000  00011ac8
  [296] __sancov_guards   PROGBITS         0000000000000000  00011acc
```

If I made the similar change to vmlinux.lds.S to module.lds.S. Then I can see
```
readelf -S /local/mnt/workspace/jiangenj/syzkaller/x86_64/linux/drivers/usb/mon/usbmon.ko|grep sancov
  [ 7] __sancov_guards   NOBITS           0000000000000000  00001460
  [52] .text.sancov[...] PROGBITS         0000000000000000  00003490
```
However, load the module, I see see
```
# insmod ./usbmon.ko
[  115.752127][ T9282] module: usbmon: Unknown rela relocation: 42
insmod: ERROR: could not insert module ./usbmon.ko: Invalid module format
```

jiangenj commented:
> > Here you go: [kcov.txt](https://github.com/user-attachments/files/18912693/kcov.txt)
> > (rename from kcov.txt to kcov.c)
> > As for modules, I haven't looked closely, but I don't expect any issues with those.
> 
> For example, If i load the module when trace_pc_guard enabled
> 
> ```
>  insmod ./usbmon.ko
> [ 3686.608008][ T9399] module: usbmon: Unknown rela relocation: 42
> insmod: ERROR: could not insert module ./usbmon.ko: Invalid module format
> ```
> 
> The readelf shows it has many __sancov_guards sections
> 
> ```
> ...
>   [294] __sancov_guards   PROGBITS         0000000000000000  00011ab8
>   [295] __sancov_guards   PROGBITS         0000000000000000  00011ac8
>   [296] __sancov_guards   PROGBITS         0000000000000000  00011acc
> ```
> 
> If I made the similar change to vmlinux.lds.S to module.lds.S. Then I can see
> 
> ```
> readelf -S /local/mnt/workspace/jiangenj/syzkaller/x86_64/linux/drivers/usb/mon/usbmon.ko|grep sancov
>   [ 7] __sancov_guards   NOBITS           0000000000000000  00001460
>   [52] .text.sancov[...] PROGBITS         0000000000000000  00003490
> ```
> 
> However, load the module, I see see
> 
> ```
> # insmod ./usbmon.ko
> [  115.752127][ T9282] module: usbmon: Unknown rela relocation: 42
> insmod: ERROR: could not insert module ./usbmon.ko: Invalid module format
> ```

Add some search results:
In llvm 42 is https://github.com/microsoft/llvm/blob/master/include/llvm/BinaryFormat/ELFRelocs/x86_64.def#L45C11-L45C33 `R_X86_64_REX_GOTPCRELX`
In objdump -r shows
```
RELOCATION RECORDS FOR [.text.sancov.module_ctor_trace_pc_guard]:
OFFSET           TYPE                     VALUE
0000000000000007 R_X86_64_REX_GOTPCRELX   __start___sancov_guards-0x4
000000000000000e R_X86_64_REX_GOTPCRELX   __stop___sancov_guards-0x4
```

ramosian-glider commented:
I am trying to investigate this. First off, I picked incorrect names for the start and end of __sancov_guards in the linker script, but fixing those is not enough.

ramosian-glider committed:
arch/x86: objtool: add support for R_X86_64_REX_GOTPCRELX

When compiling modules with -fsanitize-coverage=trace-pc-guard, Clang
will emit R_X86_64_REX_GOTPCRELX relocations for the
__start___sancov_guards and __stop___sancov_guards symbols. Although
these relocations can be resolved within the same binary, they are left
over by the linker because of the --emit-relocs flag.

This patch makes it possible to resolve the R_X86_64_REX_GOTPCRELX
relocations at runtime, as doing so does not require a .got section.
In addition, add a missing overflow check to R_X86_64_PC32/R_X86_64_PLT32.

Signed-off-by: Alexander Potapenko <glider@google.com>

ramosian-glider committed:
mm/kasan: define __asan_before_dynamic_init, __asan_after_dynamic_init

Calls to __asan_before_dynamic_init() and __asan_after_dynamic_init()
are inserted by Clang when building with coverage guards.
These functions can be used to detect initialization order fiasco bugs
in the userspace, but it is fine for them to be no-ops in the kernel.

Signed-off-by: Alexander Potapenko <glider@google.com>

jiangenj commented:
@ramosian-glider  For f57a9aa, I see this error, but no dmesg error
```
# insmod ./usbmon.ko
insmod: ERROR: could not insert module ./usbmon.ko: Invalid module format
```

ramosian-glider commented:
Could it be so that you are trying to load a stale module version?
I just loaded usbmon.ko, it works for me.

jiangenj commented:
> Could it be so that you are trying to load a stale module version? I just loaded usbmon.ko, it works for me.

still not work, can you share your build, load steps.

ramosian-glider commented:
Here is my config: 
[config.txt](https://github.com/user-attachments/files/19688570/config.txt)

I build like this: `make  KCFLAGS=-Wno-enum-enum-conversion LLVM=$CLANG_PATH/ -j72  2>&1 | tee log`

jiangenj commented:
> Here is my config: [config.txt](https://github.com/user-attachments/files/19688570/config.txt)
> 
> I build like this: `make KCFLAGS=-Wno-enum-enum-conversion LLVM=$CLANG_PATH/ -j72 2>&1 | tee log`

Thanks, my config doesn't have CONFIG_BASIC_MODVERSIONS=y which causes the insmod failure.

jiangenj commented:
@ramosian-glider can you also share the syzkaller change to align with this PR?
